src/pybind/rados.py

```diff
@@ -1667,6 +1667,39 @@ def read(self, key, length=8192, offset=0):
             raise make_ex(ret, "Ioctx.read(%s): failed to read %s" % (self.name, key))
         return ctypes.string_at(ret_buf, ret)
 
+    @requires(('key', str_type), ('cls', str_type), ('method', str_type), ('data', bytes))
+    def execute(self, key, cls, method, data, length=8192):
+        """
+        Execute an OSD class method on an object.
+
+        :param key: name of the object
+        :type key: str
+        :param cls: name of the object class
+        :type cls: str
+        :param method: name of the method
+        :type method: str
+        :param data: input data
+        :type data: bytes
+        :param length: size of output buffer in bytes (default=8291)
+        :type length: int
+
+        :raises: :class:`TypeError`
+        :raises: :class:`Error`
+        :returns: (ret, method output)
+        """
+        self.require_ioctx_open()
+        ret_buf = create_string_buffer(length)
+        ret = run_in_thread(self.librados.rados_exec,
+                (self.io, cstr(key), cstr(cls), cstr(method),
+                    c_char_p(data), c_size_t(len(data)), ret_buf,
+                    c_size_t(length)))
+        if ret < 0:
+            raise make_ex(ret, "Ioctx.exec(%s): failed to exec %s:%s on %s" %
+                    (self.name, cls, method, key))
+        # we return the exec return value as well as the full output buffer
+        # because the interpretation of these are method dependent.
+        return ret, ctypes.string_at(ret_buf, length)
```

Comment:
Yeah, that was my thinking - with this rados_exec() interface we can't tell whether we're handing the python user junk or actual output buffer, but the user can just ignore it if they're not expecting output. This assumes class method output is predictable (i.e. even with a positive return code the python user knows a priori whether to expect output). I think that's fine for now.

A rados_exec2() would be needed to break that assumption.